src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,194 @@
+  private static Integer retryCounter = 0;
+  private final ServiceNowSinkConfig config;
+  private final ServiceNowTableAPIClientImpl restApi;
+  private final Gson gson = new Gson();
+
+
+  public ServiceNowSinkAPIRequestImpl(ServiceNowSinkConfig conf) {
+    this.config = conf;
+    restApi = new ServiceNowTableAPIClientImpl(config);
+  }
+
+  public RestRequest getRestRequest(JsonObject jsonObject) {
+    JsonElement jsonElement = jsonObject.get(ServiceNowConstants.SYS_ID);
+    String sysId = null;
+
+    if (jsonElement == null) {
+      if (config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
+        throw new RuntimeException("No sys_id found in the record to be updated");
+      }
+    } else {
+      if (config.getOperation().equals(ServiceNowConstants.INSERT_OPERATION)) {
+        jsonObject.remove(ServiceNowConstants.SYS_ID);
+      }
+      sysId = jsonElement.getAsString();
+    }
+
+    String data = jsonObject.toString();
+    String encodedData = Base64.getEncoder().encodeToString(data.getBytes(StandardCharsets.UTF_8));
+
+    List<Header> headers = new ArrayList<>();
+    Header contentTypeHeader = new BasicHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
+    Header acceptHeader = new BasicHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON);
+    headers.add(contentTypeHeader);
+    headers.add(acceptHeader);
+
+    RestRequest restRequest = new RestRequest();
+    restRequest.setUrl(String.format(ServiceNowConstants.INSERT_TABLE_API_URL_TEMPLATE, config.getTableName()));
+    if (config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
+      restRequest.setUrl(String.format(ServiceNowConstants.UPDATE_TABLE_API_URL_TEMPLATE, config.getTableName(),
+        sysId));
+    }
+    restRequest.setId(counter.toString());
+    counter++;
+    restRequest.setHeaders(headers);
+    restRequest.setMethod(ServiceNowConstants.HTTP_POST);
+    if (config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
+      restRequest.setMethod(ServiceNowConstants.HTTP_PUT);
+    }
+    restRequest.setBody(encodedData);
+    return restRequest;
+  }
+
+  public Boolean createPostRequest(List<RestRequest> records) {
+    ServiceNowBatchRequest payloadRequest = getPayloadRequest(records);
+    ServiceNowTableAPIRequestBuilder requestBuilder = new ServiceNowTableAPIRequestBuilder(
+      config.getRestApiEndpoint());
+    RestAPIResponse apiResponse;
+
+    try {
+      String accessToken = restApi.getAccessToken();
+      requestBuilder.setAuthHeader(accessToken);
+      requestBuilder.setAcceptHeader(MediaType.APPLICATION_JSON);
+      requestBuilder.setContentTypeHeader(MediaType.APPLICATION_JSON);
+      StringEntity stringEntity = new StringEntity(gson.toJson(payloadRequest), ContentType.APPLICATION_JSON);
+      requestBuilder.setEntity(stringEntity);
+      apiResponse = restApi.executePost(requestBuilder.build());
+
+      if (!apiResponse.isSuccess()) {
```

Comment:
if apiResponse is a failure, how we will tackle it. if this method will return false, we will never clear restRequests and it will stop creating batches as restRequests will always be greater than default batch size.

Reply:
failing the pipeline if HTTP Status Code other than 200 is received. f52f02d4f710d02cead2bed152cda9004a27c735